File: macos/Sources/SnapikCore/Editing/SessionOperations.swift
import Foundation

/// Port of `src/Snapik.Core/Editing/SessionOperations.cs`.
public enum SessionOperations {
    public static func addCapture(_ session: SnapikSession, capture: CaptureItem, nowUtc: Date) throws -> SnapikSession {
        var updated = session
        updated.captures.append(capture)
        return try touch(updated, nowUtc)
    }

    public static func removeCapture(_ session: SnapikSession, captureId: SBGuid, nowUtc: Date) throws -> SnapikSession {
        let index = try indexOf(session.captures, captureId)
        var updated = session
        updated.captures.remove(at: index)
        return try touch(updated, nowUtc)
    }

    public static func moveCapture(
        _ session: SnapikSession,
        captureId: SBGuid,
        destinationIndex: Int,
        nowUtc: Date
    ) throws -> SnapikSession {
        guard destinationIndex >= 0 && destinationIndex < session.captures.count else {
            throw SnapikError.argumentOutOfRange("destinationIndex")
        }

        let sourceIndex = try indexOf(session.captures, captureId)
        if sourceIndex == destinationIndex {
            return session
        }

        var captures = session.captures
        let capture = captures.remove(at: sourceIndex)
        captures.insert(capture, at: destinationIndex)

        var updated = session
        updated.captures = captures
        return try touch(updated, nowUtc)
    }

    public static func updateGlobalNote(_ session: SnapikSession, note: String, nowUtc: Date) throws -> SnapikSession {
        if session.globalNote == note {
            return session
        }
        var updated = session
        updated.globalNote = note
        return try touch(updated, nowUtc)
    }

    public static func updateCapture(_ session: SnapikSession, capture: CaptureItem, nowUtc: Date) throws -> SnapikSession {
        let index = try indexOf(session.captures, capture.id)
        if session.captures[index] == capture {
            return session
        }
        var updated = session
        updated.captures[index] = capture
        return try touch(updated, nowUtc)
    }

    private static func touch(_ session: SnapikSession, _ nowUtc: Date) throws -> SnapikSession {
        var updated = session
        updated.modifiedAtUtc = nowUtc
        updated.revision += 1
        try SessionValidation.validate(updated)
        return updated
    }

    private static func indexOf(_ captures: [CaptureItem], _ captureId: SBGuid) throws -> Int {
        for (index, capture) in captures.enumerated() where capture.id == captureId {
            return index
        }
        throw SnapikError.keyNotFound("Capture \(captureId) does not exist in this session.")
    }
}

File: src/Snapik.Core/Editing/SessionOperations.cs
using System.Collections.Immutable;
using Snapik.Core.Models;

namespace Snapik.Core.Editing;

public static class SessionOperations
{
    public static SnapikSession AddCapture(SnapikSession session, CaptureItem capture, DateTimeOffset nowUtc)
    {
        ArgumentNullException.ThrowIfNull(capture);
        return Touch(session with { Captures = session.Captures.Add(capture) }, nowUtc);
    }

    public static SnapikSession RemoveCapture(SnapikSession session, Guid captureId, DateTimeOffset nowUtc)
    {
        var index = IndexOf(session.Captures, captureId);
        return Touch(session with { Captures = session.Captures.RemoveAt(index) }, nowUtc);
    }

    public static SnapikSession MoveCapture(SnapikSession session, Guid captureId, int destinationIndex, DateTimeOffset nowUtc)
    {
        if (destinationIndex < 0 || destinationIndex >= session.Captures.Length)
        {
            throw new ArgumentOutOfRangeException(nameof(destinationIndex));
        }

        var sourceIndex = IndexOf(session.Captures, captureId);
        if (sourceIndex == destinationIndex)
        {
            return session;
        }

        var capture = session.Captures[sourceIndex];
        var reordered = session.Captures.RemoveAt(sourceIndex).Insert(destinationIndex, capture);
        return Touch(session with { Captures = reordered }, nowUtc);
    }

    public static SnapikSession UpdateGlobalNote(SnapikSession session, string note, DateTimeOffset nowUtc) =>
        session.GlobalNote == note ? session : Touch(session with { GlobalNote = note }, nowUtc);

    public static SnapikSession UpdateCapture(SnapikSession session, CaptureItem capture, DateTimeOffset nowUtc)
    {
        var index = IndexOf(session.Captures, capture.Id);
        return session.Captures[index] == capture
            ? session
            : Touch(session with { Captures = session.Captures.SetItem(index, capture) }, nowUtc);
    }

    private static SnapikSession Touch(SnapikSession session, DateTimeOffset nowUtc)
    {
        var updated = session with { ModifiedAtUtc = nowUtc, Revision = checked(session.Revision + 1) };
        SessionValidation.Validate(updated);
        return updated;
    }

    private static int IndexOf(ImmutableArray<CaptureItem> captures, Guid captureId)
    {
        for (var index = 0; index < captures.Length; index++)
        {
            if (captures[index].Id == captureId)
            {
                return index;
            }
        }

        throw new KeyNotFoundException($"Capture {captureId} does not exist in this session.");
    }
}

